pag-parser2/src/nf/inference.rs

```diff
@@ -0,0 +1,234 @@
+                types.push(ty);
+            }
+            if types.len() == 1 {
+                // If there is only one field, no need to wrap in a tuple
+                Some(types.pop().unwrap())
+            } else {
+                // Otherwise, wrap in a tuple
+                Some(InferredType::Tuple(types))
+            }
+        } else {
+            // no field, unit type
+            Some(InferredType::Concrete(parse_quote! {()}))
+        }
+    }
+    /// try infer all types, but may fail with incomplete type information.
+    pub fn infer_all_types(mut self) -> HashMap<Tag, InferredType> {
+        let mut typed = 0;
+        while typed < self.nforms.len() {
+            typed = 0;
+            for i in self.nforms.keys() {
+                if self.infer(i).is_some() {
+                    typed += 1;
+                }
+            }
+        }
+        std::mem::take(self.gamma.get_mut())
+    }
+    fn infer(&self, tag: &Tag) -> Option<InferredType> {
+        match unsafe { (*self.gamma.get()).entry(tag.clone()) } {
+            // If a tag has been inferred, return its type directly
+            Entry::Occupied(entry) => Some(entry.get().clone()),
+            Entry::Vacant(slot) => Some(
+                slot.insert({
+                    // If a concrete type annotation is provided, use it directly
+                    if let Some(x) = self.annotations.get(tag) {
+                        InferredType::Concrete(x.clone())
+                    } else {
+                        let semact = self.semact.get(tag);
+                        match semact {
+                            // No semantic action, the type is unit
+                            None => InferredType::Concrete(parse_quote!(())),
+                            // Token semantic action, the type is Span
+                            Some(SemAct::Token) => {
+                                InferredType::Concrete(parse_quote!(::pag_runtime::Span<'src>))
+                            }
+                            // Customized routine without type annotation -- inference failed
+                            Some(SemAct::CustomizedRoutine(..)) => return None,
+                            // Nested routine for one or more, the type is unit.
+                            Some(SemAct::OneOrMoreNested) => {
+                                InferredType::Concrete(parse_quote!(()))
+                            }
+                            Some(SemAct::Gather) => {
+                                let nfs = self.nforms.get(tag)?;
+                                let mut inferred = None;
+                                // find first subexpression that fulfills inference
+                                for i in nfs.iter() {
+                                    let visible = i.visible_bindings(0);
+                                    if let Some(gather_type) =
+                                        self.infer_gather(visible.into_iter().map(|x| x.1))
```

Comment:
I'm afraid this can cause memory safety issues. Inside `infer_gather` you may modify the hashmap, invalidating the entry you are holding. I think you can remove `UnsafeCell` and separate this function into two phases: `gamma.contains` + `gamma.insert`